File: apps/android/app/src/main/kotlin/sh/nikhil/conduit/ui/ServerPillRow.kt
package sh.nikhil.conduit.ui

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.lazy.LazyRow
import androidx.compose.foundation.lazy.items
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import sh.nikhil.conduit.Endpoint
import sh.nikhil.conduit.HarnessState
import sh.nikhil.conduit.SavedServer

/**
 * Compose mirror of iOS `ServerPillRow` from PR #47 — a horizontally
 * scrollable strip of [ServerPill]s. Saved servers come first (curated,
 * persistent), then discovered (transient mDNS rows). Both groups are
 * keyed on the model's id (`saved:` / `discovered:` prefix), so a
 * saved+discovered pair for the same advertiser can coexist without
 * LazyRow key collisions.
 *
 * Discovery wiring is deferred to a follow-up PR. The caller supplies
 * the discovered list — current callers can pass `emptyList()` until
 * NsdManager is hoisted up out of [DiscoveryScreen].
 */
@Composable
fun ServerPillRow(
    savedServers: List<SavedServer>,
    discoveredEntries: List<DiscoveredEntry>,
    currentEndpoint: Endpoint,
    harness: HarnessState,
    onSelectSaved: (SavedServer) -> Unit,
    onSelectDiscovered: (DiscoveredEntry) -> Unit,
    modifier: Modifier = Modifier,
    /**
     * Long-press affordance on saved pills — wires through to
     * `SessionStore.forgetServer` at the call site. Discovered pills
     * still pass `null` (the `ServerPill` default), which suppresses
     * the menu. Mirror of iOS PR #128's `onForget` plumbing.
     */
    onForgetSaved: ((SavedServer) -> Unit)? = null,
) {
    val items: List<PillEntry> = buildList {
        // Saved first — the curated set should always be reachable
        // without scrolling past a transient mDNS list.
        savedServers.forEach { server ->
            val model = ServerPillModel.fromSaved(server, currentEndpoint, harness)
            add(
                PillEntry(
                    model = model,
                    onTap = { onSelectSaved(server) },
                    onForget = onForgetSaved?.let { cb -> { cb(server) } },
                ),
            )
        }
        // Then discovered. Dedupe against saved by URL: if the user has
        // already saved this advertiser we don't need to show the
        // "discovered" twin.
        val savedUrls = savedServers.map { it.endpoint.url }.toSet()
        discoveredEntries
            .filterNot { it.url in savedUrls }
            .forEach { d ->
                val model = ServerPillModel.fromDiscovered(
                    id = d.id,
                    name = d.name,
                    host = d.host,
                    port = d.port,
                    version = d.version,
                    isActive = currentEndpoint.url == d.url,
                )
                add(PillEntry(model = model, onTap = { onSelectDiscovered(d) }, onForget = null))
            }
    }

    LazyRow(
        modifier = modifier,
        horizontalArrangement = Arrangement.spacedBy(8.dp),
        contentPadding = PaddingValues(horizontal = 16.dp),
    ) {
        items(items, key = { it.model.id }) { entry ->
            ServerPill(model = entry.model, onTap = entry.onTap, onForget = entry.onForget)
        }
    }
}

private data class PillEntry(
    val model: ServerPillModel,
    val onTap: () -> Unit,
    val onForget: (() -> Unit)?,
)

/**
 * Lightweight transport for a discovered LAN server. Lives separately
 * from [ServerPillModel] because the model is intentionally pure
 * (no token, no URL — just display fields), while callers need the
 * full pairing data to actually connect.
 *
 * Source: the NsdManager browser inside [DiscoveryScreen]. A future PR
 * hoists that browser into the store so the Home strip can stream rows
 * in too; until then [discoveredEntries] is wired as `emptyList()` from
 * Home callers.
 */
data class DiscoveredEntry(
    val id: String,
    val name: String,
    val host: String,
    val port: Int,
    val token: String,
    val version: String?,
) {
    /** Convenience: the ws:// URL the pairing would use. */
    val url: String
        get() {
            val hostStr = if (host.contains(':')) "[$host]" else host
            return "ws://$hostStr:$port"
        }
}

File: apps/ios/Sources/Shared/ServerPillRow.swift
import SwiftUI

/// Horizontal scrollable strip of `ServerPill` — saved first, then
/// discovered, then trailing `+ server` CTA. Used both on Home and
/// inside the DiscoveryView so the litter-style "pill row" affordance
/// stays consistent across the two surfaces.
///
/// Discovery state is provided by the caller (an `@State` browser),
/// not owned by this view. That keeps the row reusable: HomeView can
/// hold a long-lived browser; DiscoveryView can show the same row as
/// a "what you'll see on Home" preview.
struct ServerPillRow: View {
    @Environment(SessionStore.self) private var store
    /// Discovered rows piped in from a `LANDiscoveryBrowser` (or any
    /// other source). Empty when discovery is off.
    var discovered: [LANDiscoveryBrowser.Discovered] = []
    @Binding var showAddServer: Bool

    var body: some View {
        ScrollView(.horizontal, showsIndicators: false) {
            HStack(spacing: 8) {
                ForEach(savedModels) { model in
                    ServerPill(
                        model: model,
                        onTap: { selectSaved(model) },
                        onForget: { forgetSaved(model) },
                        onRename: { renameSaved(model) }
                    )
                }
                ForEach(discoveredModels) { model in
                    ServerPill(
                        model: model,
                        onTap: { connectDiscovered(model) },
                        onAdopt: { connectDiscovered(model) }
                    )
                }
                addPill
            }
            .padding(.horizontal, 14)
            .padding(.vertical, 4)
        }
    }

    // MARK: - Models

    private var savedModels: [ServerPillModel] {
        store.savedServers.map {
            ServerPillModel.fromSaved($0, currentEndpoint: store.endpoint, harness: store.harness)
        }
    }

    /// Discovered rows minus any whose host:port already appears in the
    /// saved list — surfacing the same advertiser twice would just be
    /// noise (and pairing flow naturally adopts it into `savedServers`).
    private var discoveredModels: [ServerPillModel] {
        let savedKeys: Set<String> = Set(store.savedServers.compactMap { savedKey($0) })
        return discovered.compactMap { row in
            let key = "\(row.host):\(row.port)"
            if savedKeys.contains(key) { return nil }
            return ServerPillModel.fromDiscovered(
                id: row.id,
                name: row.name,
                host: row.host,
                port: row.port,
                version: row.version,
                isActive: false
            )
        }
    }

    private func savedKey(_ server: SavedServer) -> String? {
        guard let (host, port) = ServerPillModel.splitHostPort(server.endpoint.url) else { return nil }
        return "\(host):\(port)"
    }

    // MARK: - Actions

    private func selectSaved(_ model: ServerPillModel) {
        let savedID = String(model.id.dropFirst("saved:".count))
        store.selectSavedServer(savedID, autoConnect: true)
    }

    private func forgetSaved(_ model: ServerPillModel) {
        let savedID = String(model.id.dropFirst("saved:".count))
        store.forgetServer(savedID)
    }

    private func renameSaved(_ model: ServerPillModel) {
        // The actual rename UI lives in SettingsSheet; the
        // context-menu hook in v1 of this row is a no-op placeholder
        // — wiring up an alert/sheet from inside a horizontal-scroll
        // cell is out of scope for the discovery-flesh-out PR. The
        // hook stays so a later pass can drop in an alert presenter
        // without churning the call sites.
        _ = model
    }

    private func connectDiscovered(_ model: ServerPillModel) {
        guard let row = discovered.first(where: { "discovered:\($0.id)" == model.id }) else { return }
        let endpoint = StoredEndpoint(
            url: "ws://\(row.host):\(row.port)",
            token: row.token
        )
        store.endpoint = endpoint
        store.upsertSavedServer(name: row.name, endpoint: endpoint, makeDefault: true)
        store.disconnect()
        store.connect()
    }

    private var addPill: some View {
        Button {
            showAddServer = true
        } label: {
            HStack(spacing: 6) {
                Image(systemName: "plus")
                    .font(.caption.weight(.bold))
                Text("server")
                    .font(.subheadline.weight(.semibold))
            }
            .foregroundStyle(ConduitTheme.accentStrong)
            .padding(.horizontal, 12)
            .padding(.vertical, 8)
            .background(
                Capsule()
                    .stroke(ConduitTheme.accentStrong, lineWidth: 1.5)
            )
        }
        .buttonStyle(.plain)
        .accessibilityLabel("Add server")
    }
}

/// Pure-data merge policy for the `ServerPillRow`. Lifted out so the
/// dedupe-by-host:port rule and the saved-first ordering can be pinned
/// without a SwiftUI host. Mirrors the Rust `DiscoveryRegistry`'s
/// merge semantics (Saved wins over Discovered when host:port matches)
/// at the iOS layer — the registry will eventually own this, but until
/// the UniFFI surface lands we keep the policy here so the iOS view
/// has somewhere to point its tests at.
enum DiscoveryMergePolicy {
    struct DiscoveredInput: Equatable {
        let id: String
        let name: String
        let host: String
        let port: Int
        let version: String?
    }

    /// Merge saved + discovered into the ordered list of pills the row
    /// will render. Saved pills come first; discovered pills are
    /// filtered to remove anything whose host:port already appears in
    /// the saved set.
    static func merge(
        saved: [SavedServer],
        discovered: [DiscoveredInput],
        currentEndpoint: StoredEndpoint,
        harness: HarnessState
    ) -> [ServerPillModel] {
        let savedModels = saved.map {
            ServerPillModel.fromSaved($0, currentEndpoint: currentEndpoint, harness: harness)
        }
        let savedKeys: Set<String> = Set(saved.compactMap { server in
            guard let (host, port) = ServerPillModel.splitHostPort(server.endpoint.url) else { return nil }
            return "\(host):\(port)"
        })
        let discoveredModels = discovered.compactMap { input -> ServerPillModel? in
            let key = "\(input.host):\(input.port)"
            if savedKeys.contains(key) { return nil }
            return ServerPillModel.fromDiscovered(
                id: input.id,
                name: input.name,
                host: input.host,
                port: input.port,
                version: input.version,
                isActive: false
            )
        }
        return savedModels + discoveredModels
    }
}
